default/AI/ResearchAI.py

```diff
@@ -1,30 +1,328 @@
+    # further specialization for organic hulls
+    orgneural = hull if org == hull and index & 8 == 0 else offtrack_subhull
+    orgraven = hull if org == hull and index & 8 == 1 else offtrack_subhull
+    orgendosym = hull if org == hull and (index + 17) % 43 < 22 else offtrack_subhull
+    # further specialization for energy hulls
+    energyfrac = hull if energy == hull and index & 8 == 0 else offtrack_subhull
+    energymag = hull if energy == hull and index & 8 == 1 else offtrack_subhull
+    # repair techs may be skipped if AI decides to go for nanorobotic hull which full-repairs
+    repair = 1 if not nanorobo == hull or (index + 37) % 103 < 77 else 0.3
+
+    # AI may skip weapon lines
+    weapon = 1
+    massdriver = weapon if index % 47 < 36 else 0
+    laser = weapon if (index % 83 < 36 if massdriver == weapon else index % 47 < 41) else 0
+    plasmacannon = weapon if (index % 83 < 36 if laser == weapon else index % 47 < 41) else 0
+    deathray = weapon
+
+    armor = 1
+
+    engine = 1 if ((index + 13) % 31 < 26 if ColonisationAI.galaxy_is_sparse() else (index + 13) % 31 < 5) else 0
+    fuel = 1 if ((index + 17) % 37 < 31 if ColonisationAI.galaxy_is_sparse() else (index + 17) % 37 < 3) else 0
+
+    detection = 1
+    stealth = 2.2 if (index + 37) % 67 < 9 else 0.2 # TODO stealthy species probably want more stealthy techs
+
+    if tech_name in ["SHP_KRILL_SPAWN", "DEF_PLANET_CLOAK"]:
+        return unresearchable
+
+    # defense techs
+    if tech_name in ["DEF_GARRISON_1", "DEF_GARRISON_2", "DEF_GARRISON_3", "DEF_GARRISON_4"]:
+        return defensive - 0.01 # garrison are less important/powerful among all defenses
+    if tech_name in ["DEF_DEFENSE_NET_1", "DEF_DEFENSE_NET_2", "DEF_DEFENSE_NET_3"]:
+        return defensive
+    if tech_name in ["DEF_DEFENSE_NET_REGEN_1", "DEF_DEFENSE_NET_REGEN_2"]:
+        return defensive
+    if tech_name in ["DEF_PLAN_BARRIER_SHLD_1", "DEF_PLAN_BARRIER_SHLD_2", "DEF_PLAN_BARRIER_SHLD_3", "DEF_PLAN_BARRIER_SHLD_4", "DEF_PLAN_BARRIER_SHLD_5"]:
+        return defensive
+    if tech_name in ["DEF_SYST_DEF_MINE_1", "DEF_SYST_DEF_MINE_2", "DEF_SYST_DEF_MINE_3"]:
+        return defensive + 0.01 # mines are more important/powerful among all defenses
+
+    # learning
+    if tech_name in ["LRN_PHYS_BRAIN", "LRN_TRANSLING_THT", "LRN_PSIONICS", "LRN_GRAVITONICS", "LRN_EVERYTHING", "LRN_MIND_VOID", "LRN_NDIM_SUBSPACE", "LRN_TIME_MECH"]:
+        return theory
+    if tech_name in ["LRN_ALGO_ELEGANCE", "LRN_ARTIF_MINDS", "LRN_DISTRIB_THOUGHT", "LRN_QUANT_NET", "LRN_STELLAR_TOMOGRAPHY"]:
+        return research_boost
+    if tech_name in ["LRN_ENCLAVE_VOID", "LRN_UNIF_CONC"]:
+        return research_boost
+    if tech_name in ["LRN_XENOARCH"]:
+        if foAI.foAIstate.aggression < fo.aggression.typical:
+            return default
+        return immediate if ColonisationAI.gotRuins else useless # get xenoarcheology ASAP when we have ruins, otherwise it is useless
```

Comment:
If we happen to catch ruins within 2 turns with our starting ship, this will severely cripple our development as the tech cost is relatively high and the RP better spent elsewhere. Some more conditions are probably a good idea.